crypto/ring-signature/src/ring_signature/mlsag_sign.rs

```diff
@@ -0,0 +1,343 @@
+
+        // Iterate around the ring, starting at real_index.
+        // NOTE: THIS REORDERING IS CRITICAL FOR PIECEWISE COMPUTATION
+        for n in 0..ring_size {
+            let i = (self.real_index + n) % ring_size;
+            let tx_out = &ring.index(i)?;
+
+            sign_ctx.update(&self, i, tx_out)?;
+        }
+
+        // "Close the loop" by computing responses for the real index.
+        let (key_image, c_zero) = sign_ctx.finalise(&self)?;
+
+        Ok((key_image, c_zero))
+    }
+}
+
+/// MLSAG signing context, supports incremental computation for ring signing.
+///
+/// WARNING: MISUSE OF THIS API MAY GENERATE INVALID RINGS,
+/// SEE [`MlsagSign`] FOR A PREFERRED HIGHER LEVEL ABSTRACTION
+#[derive(Debug, Zeroize)]
+pub struct MlsagSignCtx<R: AsMut<[CurveScalar]>> {
+    alpha_0: Scalar,
+    alpha_1: Scalar,
+    G: RistrettoPoint,
+    I: RistrettoPoint,
+
+    /// Cache real input for balance checking
+    #[zeroize(skip)]
+    real_input: Option<(RistrettoPublic, Commitment)>,
+
+    /// Number of rings computed
+    ring_count: usize,
+
+    responses: R,
+
+    real_challenge: Option<Scalar>,
+    last_challenge: Option<Scalar>,
+    zeroth_challenge: Option<Scalar>,
+
+    #[zeroize(skip)]
+    key_image: KeyImage,
+
+    #[zeroize(skip)]
+    output_commitment: Commitment,
+
+    complete: bool,
+}
+
+#[allow(dead_code)]
+impl<R: AsRef<[CurveScalar]> + AsMut<[CurveScalar]>> MlsagSignCtx<R> {
+    /// Initialise signing state, including `challenges` and `responses` working
+    /// buffers
+    pub fn init<'a>(
+        params: &'a MlsagSignParams<'a>,
+        mut rng: impl CryptoRngCore,
+        mut responses: R,
+    ) -> Result<Self, Error> {
+        let G = B_BLINDING;
+        debug_assert!(
+            params.generator.B_blinding == G,
+            "basepoint for blindings mismatch"
+        );
+
+        let r = responses.as_mut();
+
+        // Check buffer lengths are correct for ring size
+        if r.len() != params.ring_size * 2 {
+            return Err(Error::LengthMismatch(params.ring_size * 2, r.len()));
+        }
+
+        // Generate KeyImage from generated transaction private key
+        let key_image = KeyImage::from(params.onetime_private_key);
+
+        // The uncompressed key_image.
+        let I = key_image.point.decompress().ok_or(Error::InvalidKeyImage)?;
+
+        // Uncompressed output commitment.
+        // This ensures that each address and commitment encodes a valid Ristretto
+        // point.
+        let output_commitment =
+            Commitment::new(params.value, *params.output_blinding, params.generator);
+
+        // Responses `r_{0,0}, r_{0,1}, ... , r_{ring_size-1,0}, r_{ring_size-1,1}`.
+        r.iter_mut()
+            .for_each(|v| *v = CurveScalar::from(Scalar::zero()));
+
+        for i in 0..params.ring_size {
+            if i == params.real_index {
+                continue;
+            }
+            r[2 * i].scalar = Scalar::random(&mut rng);
```

Comment:
i found an issue with a hw RNG because all but the real response were being set to zero... would it be reasonable to add a check / retry / fail to make sure this can never happen in reality?

Reply:
I think such check should be in the implementation of the object that implements CryptoRng

Reply:
Compare with: https://github.com/mobilecoinfoundation/mobilecoin/blob/master/crypto/rand/src/rdrandrng/retry.rs

Reply:
hmm yeah that's probably fair. as a corollary, are the properties of the ring maintained _if_ one or more responses are zero?

Reply:
the conclusion was, no this is definitely a problem. added an RNG check to the ledger app to mitigate.